src/transforms/dedupe.rs

```diff
@@ -0,0 +1,269 @@
+mod tests {
+    use super::Dedupe;
+    use crate::{event::Event, transforms::Transform};
+
+    #[test]
+    fn dedupe_test_basic_matching() {
+        let mut event1 = Event::from("message");
+        event1.as_mut_log().insert("matched", "some value");
+        event1.as_mut_log().insert("unmatched", "another value");
+
+        // Test that unmatched field isn't considered
+        let mut event2 = Event::from("message");
+        event2.as_mut_log().insert("matched", "some value2");
+        event2.as_mut_log().insert("unmatched", "another value");
+
+        // Test that matched field is considered
+        let mut event3 = Event::from("message");
+        event3.as_mut_log().insert("matched", "some value");
+        event3.as_mut_log().insert("unmatched", "another value2");
+
+        let mut transform = Dedupe::new(5, vec!["matched".into()]);
+
+        // First event should always be passed through as-is.
+        let new_event = transform.transform(event1).unwrap();
+        assert_eq!(new_event.as_log()[&"matched".into()], "some value".into());
+
+        // Second event differs in matched field so should be outputted even though it
+        // has the same value for unmatched field.
+        let new_event = transform.transform(event2).unwrap();
+        assert_eq!(new_event.as_log()[&"matched".into()], "some value2".into());
+
+        // Third event has the same value for "matched" as first event, so it should be dropped.
+        assert_eq!(None, transform.transform(event3));
+    }
+
+    #[test]
+    fn dedupe_test_field_name_matters() {
+        let mut event1 = Event::from("message");
+        event1.as_mut_log().insert("matched1", "some value");
+
+        let mut event2 = Event::from("message");
+        event2.as_mut_log().insert("matched2", "some value");
+
+        let mut transform = Dedupe::new(5, vec!["matched1".into(), "matched2".into()]);
+
+        // First event should always be passed through as-is.
+        let new_event = transform.transform(event1).unwrap();
+        assert_eq!(new_event.as_log()[&"matched1".into()], "some value".into());
+
+        // Second event has a different matched field name with the same value, so it should not be
+        // considered a dupe
+        let new_event = transform.transform(event2).unwrap();
+        assert_eq!(new_event.as_log()[&"matched2".into()], "some value".into());
+    }
+
+    /**
+     * Test that two Events that are considered duplicates get handled that way, even
+     * if the order of the matched fields is different between the two.
+     * Note that this test is a bit non-deterministic in its ability to test the underlying
+     * issue (meaning it's possible for this test to pass even if the bug of the transform
+     * caring about field order is present).  This is difficult to avoid because Event is
+     * backed by a HashMap and there's no way to deterministically control what order the
+     * objects will be stored in table underlying the HashMap.
```

Comment:
So https://github.com/timberio/vector/pull/1838 would fix this?

Reply:
Ooh yes.  It would also simplify the patch I'm working on now which is a follow-up to this one that allows users to specify the fields to ignore rather than fields to match.